File: src/main/java/rails/game/action/PossibleActions.java
package rails.game.action;

import java.util.EnumSet;
import java.util.List;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.Lists;

/**
 * This class manages the actions that the current user can execute at any point
 * in time. Each possible action is represented by an instance of a subclass of
 * PossibleAction. The complete set is stored in an ArrayList.
 * 
 * TODO: Should this be changed to a set?
 */
public class PossibleActions {

    private final List<PossibleAction> actions = Lists.newArrayList();

    private PossibleActions() { }

    public static PossibleActions create() {
        return new PossibleActions();
    }

    public void clear() {
        actions.clear();
    }

    public void add(PossibleAction action) {
        actions.add(action);
    }

    public void remove(PossibleAction action) {
        actions.remove(action);
    }

    public void addAll(List<? extends PossibleAction> actions) {
        this.actions.addAll(actions);
    }

    public boolean contains(Class<? extends PossibleAction> clazz) {
        for (PossibleAction action : actions) {
            if (clazz.isAssignableFrom(action.getClass())) return true;
        }
        return false;
    }

    @SuppressWarnings("unchecked")
    public <T extends PossibleAction> ImmutableList<T> getType(Class<T> clazz) {
        ImmutableList.Builder<T> result = ImmutableList.builder();
        for (PossibleAction action : actions) {
            if (clazz.isAssignableFrom(action.getClass())) {
                result.add((T) action);
            }
        }
        return result.build();
    }

    public ImmutableList<PossibleAction> getList() {
        return ImmutableList.copyOf(actions);
    }

    public boolean isEmpty() {
        return actions.isEmpty();
    }

    public boolean containsOnlyPass() {
        if (actions.size() != 1) return false;
        PossibleAction action = actions.get(0);
        if (action instanceof NullAction && ((NullAction)action).getMode() == NullAction.Mode.PASS) {
            return true;
        } else {
            return false;
        }
    }
    
    public boolean containsCorrections() {
        for (PossibleAction action:actions) {
            if (action.isCorrection()) return true;
        }
        return false;
    }

    /** Check if a given action exists in the current list of possible actions */
    public boolean validate(PossibleAction checkedAction) {

        // Some actions are always allowed
        if (checkedAction instanceof GameAction
                && EnumSet.of(GameAction.Mode.SAVE, GameAction.Mode.RELOAD, GameAction.Mode.EXPORT).contains(
                        ((GameAction)checkedAction).getMode() )) {
            return true;
        }

        // Check if action accurs in the list of possible actions
        for (PossibleAction action : actions) {
            if (action.equalsAsOption(checkedAction)) {
                return true;
            }
        }
        return false;
    }

}

File: dotnet/RailsLib.Net/Rails/Game/Action/PossibleActions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
/**
 * This class manages the actions that the current user can execute at any point
 * in time. Each possible action is represented by an instance of a subclass of
 * PossibleAction. The complete set is stored in a List.
 * 
 * TODO: Should this be changed to a set?
 */
namespace GameLib.Rails.Game.Action
{
    public class PossibleActions
    {
        private readonly List<PossibleAction> actions;

        private PossibleActions()
        {
            actions = new List<PossibleAction>();
        }

        public static PossibleActions Create()
        {
            return new PossibleActions();
        }

        public void Clear()
        {
            actions.Clear();
        }

        public void Add(PossibleAction action)
        {
            actions.Add(action);
        }

        public void Remove(PossibleAction action)
        {
            actions.Remove(action);
        }

        public void AddAll<T>(List<T> actions) where T : PossibleAction
        {
            this.actions.AddRange(actions);
        }

        public bool Contains(Type type)
        {
            foreach (PossibleAction action in actions)
            {
                if (type.IsAssignableFrom(action.GetType())) return true;
            }
            return false;
        }

        // Removed, because I think the one below is good enough
        //public List<T> GetActionType<T>(Type type) where T : PossibleAction
        //{
        //    List<T> result = new List<T>();
        //    foreach (PossibleAction action in actions)
        //    {
        //        if (type.IsAssignableFrom(action.GetType()))
        //        {
        //            result.Add((T)action);
        //        }
        //    }
        //    return result;
        //}

        public List<T> GetActionType<T>() where T : PossibleAction
        {
            List<T> result = new List<T>();
            foreach (PossibleAction action in actions)
            {
                if (typeof(T).IsAssignableFrom(action.GetType()))
                {
                    result.Add((T)action);
                }
            }
            return result;
        }

        public List<PossibleAction> GetList()
        {
            return new List<PossibleAction>(actions); //ImmutableList.copyOf(actions);
        }

        public bool IsEmpty
        {
            get
            {
                return actions.Count == 0;
            }
        }

        public bool ContainsOnlyPass()
        {
            if (actions.Count != 1) return false;
            PossibleAction action = actions[0];
            if (action is NullAction && ((NullAction)action).Mode == NullAction.Modes.PASS)
            {
                return true;
            }
            else
            {
                return false;
            }
        }

        public bool ContainsCorrections()
        {
            foreach (PossibleAction action in actions)
            {
                if (action.IsCorrection) return true;
            }
            return false;
        }

        /** Check if a given action exists in the current list of possible actions */
        public bool Validate(PossibleAction checkedAction)
        {

            // Some actions are always allowed
            var checkList = new List<GameAction.Modes> { GameAction.Modes.SAVE, GameAction.Modes.RELOAD, GameAction.Modes.EXPORT };
            if (checkedAction is GameAction && checkList.Contains(((GameAction)checkedAction).Mode))
            {
                return true;
            }

            // Check if action occurs in the list of possible actions
            foreach (PossibleAction action in actions)
            {
                if (action.EqualsAsOption(checkedAction))
                {
                    return true;
                }
            }
            return false;
        }
    }
}
